Fix bug involving defaults

crcornwell opened the pull request:
Now we'll only set defaults on an update schema if the fields are linked to the updated field.

Commit:
FBCM-1897 don't set defaults on updateSchemas unless they are linked fields

davezuch commented on src/cn-batch-forms.service.js:
```diff
@@ -157,6 +162,24 @@
 
+    function processDiff(schema) {
+      const updateSchema = schema.params.updateSchema;
+      const links = _.filter(schema.batchConfig.links, ls => _.includes(ls, updateSchema));
```
I would switch from `_.includes` to `_.startsWith`; I ran into a bug where I was doing something similar and when the `updateSchema` value was `creative` it would match the key `admin.creative_name` which was unrelated and we didn't want

davezuch commented on src/cn-batch-forms.service.js:
```diff
@@ -157,6 +162,24 @@
+    function processDiff(schema) {
+      const updateSchema = schema.params.updateSchema;
+      const links = _.filter(schema.batchConfig.links, ls => _.includes(ls, updateSchema));
+      const hardLinks = _.filter(schema.batchConfig.hardLinks, ls => _.includes(ls, updateSchema));
```
Same ^

davezuch commented on src/cn-batch-forms.service.js:
```diff
@@ -157,6 +162,24 @@
+    function processSchemaDiff(properties, links) {
+      const props = _.keys(properties);
+      _.forEach(props, (prop) => {
+        if (_.has(properties[prop], "properties")) {
```
This handles objects but I don't see a case for arrays. If a schema type is `object` it'll contain the `properties` prop, but if it's `array` it'll contain an `items` prop that contains the array child's schema

davezuch commented on src/cn-batch-forms.service.js:
```diff
@@ -157,6 +162,24 @@
+      _.forEach(props, (prop) => {
+        if (_.has(properties[prop], "properties")) {
+          processSchemaDiff.call(this, properties[prop].properties, links);
+        } else if (_.every(links, l => !_.includes(l, prop))) {
```
I'm not sure if this `_.includes` also needs to be changed to `_.startsWith`. This would be a great place to add a test to make sure we get the desired results for different possible scenarios like the one I mentioned above, or for when processing an array schema

Commit:
review: address notes from pull review
make sure processSchemaDiff properly handles array types
use startsWith instead of includes to prevent false positives on similar keys

Commit:
chore: add webpack, gulp and tape

Commit:
chore: add testing tasks for gulpfile

Commit:
chore: use webpack for the build

Commit:
chore: remove gulpfile and gulp dependencies

Commit:
chore: add test for clearSchemaDefault and corresponding refactor

Commit:
chore: add tests and refactor to fix unwanted behavior

Commit:
chore: add test for stripping defaults on similar keys in processDiff

Commit:
fix: remove debug statements

davezuch commented:
![high five](http://www.reactiongifs.com/r/tghf.gif)
